File: frontend/src/api/games.js
/**
 * API service for games and rounds
 * Connects to the FastAPI backend
 */

const API_BASE_URL = import.meta.env.VITE_API_BASE_URL || "http://localhost:8000";

/**
 * Create or get a game
 * @param {Object} gameData - {code?, starting_cash, status}
 * @returns {Promise<{success: boolean, game?: Object, error?: string}>}
 */
export async function createOrGetGame(gameData = {}) {
  try {
    const response = await fetch(`${API_BASE_URL}/api/games`, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify({
        code: gameData.code || null,
        starting_cash: gameData.starting_cash || 10000,
        status: gameData.status || "active",
      }),
    });
    
    if (!response.ok) {
      throw new Error(`Failed to create/get game: ${response.status} ${response.statusText}`);
    }
    
    const data = await response.json();
    return { success: true, game: data.game };
  } catch (error) {
    console.error("Error creating/getting game:", error);
    return { success: false, error: error.message, game: null };
  }
}

/**
 * Get a game by ID
 * @param {number} gameId
 * @returns {Promise<{success: boolean, game?: Object, error?: string}>}
 */
export async function getGameById(gameId) {
  try {
    const response = await fetch(`${API_BASE_URL}/api/games/${gameId}`, {
      method: "GET",
      headers: {
        "Content-Type": "application/json",
      },
    });
    
    if (!response.ok) {
      if (response.status === 404) {
        return { success: false, error: "Game not found", game: null };
      }
      throw new Error(`Failed to fetch game: ${response.status} ${response.statusText}`);
    }
    
    const data = await response.json();
    return { success: true, game: data.game };
  } catch (error) {
    console.error("Error fetching game:", error);
    return { success: false, error: error.message, game: null };
  }
}

/**
 * Create or get a round
 * @param {Object} roundData - {game_id, round_no}
 * @returns {Promise<{success: boolean, round?: Object, error?: string}>}
 */
export async function createOrGetRound(roundData) {
  try {
    const response = await fetch(`${API_BASE_URL}/api/games/rounds`, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify({
        game_id: roundData.game_id,
        round_no: roundData.round_no,
      }),
    });
    
    if (!response.ok) {
      throw new Error(`Failed to create/get round: ${response.status} ${response.statusText}`);
    }
    
    const data = await response.json();
    return { success: true, round: data.round };
  } catch (error) {
    console.error("Error creating/getting round:", error);
    return { success: false, error: error.message, round: null };
  }
}

/**
 * Get a round by ID
 * @param {number} roundId
 * @returns {Promise<{success: boolean, round?: Object, error?: string}>}
 */
export async function getRoundById(roundId) {
  try {
    const response = await fetch(`${API_BASE_URL}/api/games/rounds/${roundId}`, {
      method: "GET",
      headers: {
        "Content-Type": "application/json",
      },
    });
    
    if (!response.ok) {
      if (response.status === 404) {
        return { success: false, error: "Round not found", round: null };
      }
      throw new Error(`Failed to fetch round: ${response.status} ${response.statusText}`);
    }
    
    const data = await response.json();
    return { success: true, round: data.round };
  } catch (error) {
    console.error("Error fetching round:", error);
    return { success: false, error: error.message, round: null };
  }
}

/**
 * End a round
 * @param {number} roundId
 * @returns {Promise<{success: boolean, round?: Object, error?: string}>}
 */
export async function endRound(roundId) {
  try {
    const response = await fetch(`${API_BASE_URL}/api/games/rounds/${roundId}/end`, {
      method: "PUT",
      headers: {
        "Content-Type": "application/json",
      },
    });
    
    if (!response.ok) {
      if (response.status === 404) {
        return { success: false, error: "Round not found", round: null };
      }
      throw new Error(`Failed to end round: ${response.status} ${response.statusText}`);
    }
    
    const data = await response.json();
    return { success: true, round: data.round };
  } catch (error) {
    console.error("Error ending round:", error);
    return { success: false, error: error.message, round: null };
  }
}

File: backend/routers/games.py
"""
API routes for games and rounds.
"""
from fastapi import APIRouter, HTTPException
from backend.models import Game, GameCreate, GameResponse, Round, RoundCreate, RoundResponse
from backend.services import game_service

router = APIRouter(prefix="/api/games", tags=["games"])


@router.post("", response_model=GameResponse, status_code=201)
async def create_or_get_game(game_data: GameCreate):
    """
    Create a new game or get existing game by code.
    
    - **code**: Optional game code (for multiplayer games)
    - **starting_cash**: Starting cash amount (default: 10000)
    - **status**: Game status (default: "active")
    """
    try:
        game = game_service.create_or_get_game(
            code=game_data.code,
            starting_cash=game_data.starting_cash,
            status=game_data.status
        )
        
        return GameResponse(
            success=True,
            game=game,
            message="Game created or retrieved successfully"
        )
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Error creating/getting game: {str(e)}")


@router.get("/{game_id}", response_model=GameResponse)
async def get_game_by_id(game_id: int):
    """
    Get a game by its ID.
    
    - **game_id**: The game ID
    """
    game = game_service.get_game_by_id(game_id)
    if not game:
        raise HTTPException(status_code=404, detail=f"Game with id '{game_id}' not found")
    
    return GameResponse(
        success=True,
        game=game,
        message="Game retrieved successfully"
    )


@router.post("/rounds", response_model=RoundResponse, status_code=201)
async def create_or_get_round(round_data: RoundCreate):
    """
    Create a new round or get existing round for the game and round number.
    
    - **game_id**: The game ID
    - **round_no**: The round number (1, 2, 3...)
    """
    try:
        round_obj = game_service.create_or_get_round(
            game_id=round_data.game_id,
            round_no=round_data.round_no
        )
        
        return RoundResponse(
            success=True,
            round=round_obj,
            message="Round created or retrieved successfully"
        )
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Error creating/getting round: {str(e)}")


@router.get("/rounds/{round_id}", response_model=RoundResponse)
async def get_round_by_id(round_id: int):
    """
    Get a round by its ID.
    
    - **round_id**: The round ID
    """
    round_obj = game_service.get_round_by_id(round_id)
    if not round_obj:
        raise HTTPException(status_code=404, detail=f"Round with id '{round_id}' not found")
    
    return RoundResponse(
        success=True,
        round=round_obj,
        message="Round retrieved successfully"
    )


@router.put("/rounds/{round_id}/end", response_model=RoundResponse)
async def end_round(round_id: int):
    """
    Mark a round as ended by setting ends_at timestamp.
    
    - **round_id**: The round ID
    """
    round_obj = game_service.end_round(round_id)
    if not round_obj:
        raise HTTPException(status_code=404, detail=f"Round with id '{round_id}' not found")
    
    return RoundResponse(
        success=True,
        round=round_obj,
        message="Round ended successfully"
    )

